SingularityService/src/main/java/com/hubspot/singularity/data/LoggingCuratorFramework.java

```diff
@@ -0,0 +1,319 @@
+import org.apache.curator.framework.api.transaction.CuratorMultiTransaction;
+import org.apache.curator.framework.api.transaction.CuratorTransaction;
+import org.apache.curator.framework.api.transaction.TransactionOp;
+import org.apache.curator.framework.imps.CuratorFrameworkState;
+import org.apache.curator.framework.listen.Listenable;
+import org.apache.curator.framework.schema.SchemaSet;
+import org.apache.curator.framework.state.ConnectionStateErrorPolicy;
+import org.apache.curator.framework.state.ConnectionStateListener;
+import org.apache.curator.utils.EnsurePath;
+import org.apache.zookeeper.Watcher;
+import org.apache.zookeeper.server.quorum.flexible.QuorumVerifier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LoggingCuratorFramework implements CuratorFramework {
+  private final CuratorFramework curator;
+
+  private final ConcurrentMap<String, Integer> counters;
+
+  private final Long interval = 3L;
+
+  private static final Logger LOG = LoggerFactory.getLogger(
+    LoggingCuratorFramework.class
+  );
+
+  private final ScheduledExecutorService executorService;
+
+  private class ClearCounter implements Runnable {
+
+    @Override
+    public void run() {
+      logAndClear();
+    }
+  }
+
+  @Inject
+  public LoggingCuratorFramework(
+    CuratorFramework curator,
+    SingularityManagedScheduledExecutorServiceFactory executorServiceFactory
+  ) {
+    this.curator = curator;
+    counters = new ConcurrentHashMap<>();
+
+    this.executorService = executorServiceFactory.get("logging-curator-framework");
+    this.executorService.scheduleAtFixedRate(
+        new ClearCounter(),
+        interval,
+        interval,
+        TimeUnit.MINUTES
+      );
+  }
+
+  private String getCaller() {
+    StackTraceElement[] stackTraceElements = Thread.currentThread().getStackTrace();
+    long threadId = Thread.currentThread().getId();
+
+    int levelInStack = 0;
+    String className = "";
+    for (int i = 0; i < stackTraceElements.length; i++) {
+      String longClassName = stackTraceElements[i].getClassName();
+      className = longClassName.substring(longClassName.lastIndexOf(".") + 1);
+
+      if (
+        !(
+          className.equals("Thread") ||
+          className.equals("LoggingCuratorFramework") ||
+          className.equals("CuratorManager") ||
+          className.equals("CuratorAsyncManager")
+        )
+      ) {
+        levelInStack = i;
+        break;
+      }
+    }
+
+    return (
+      threadId +
+      ": " +
+      className +
+      "\'s " +
+      stackTraceElements[levelInStack].getMethodName()
+    );
+  }
+
+  private void incCounter() {
+    String caller = getCaller();
+    counters.compute(caller, (k, v) -> v == null ? 1 : v + 1);
+  }
+
+  public void logAndClear() {
+    counters.forEach(
+      (caller, count) -> {
+        LOG.info("{} called ZK {} times in {} minutes", caller, count, interval);
+
+        counters.remove(caller);
+      }
+    );
+  }
```

Comment:
Think I missed this before — doing `counters.put(caller, 0)` instead of `remove` would prevent any iterator invalidation stemming from removal (or additions from other threads) during iteration.